src/util/lp/rooted_mons.h

```diff
@@ -0,0 +1,239 @@
+namespace nla {
+struct index_with_sign {
+    unsigned m_i; // the index
+    rational m_sign; // the sign: -1 or 1
+    index_with_sign(unsigned i, rational sign) : m_i(i), m_sign(sign) {}
+    index_with_sign() {}
+    bool operator==(const index_with_sign& b) {
+        return m_i == b.m_i && m_sign == b.m_sign;
+    }
+    unsigned var() const { return m_i; }
+    unsigned index() const { return m_i; }
+    const rational& sign() const { return m_sign; }
+    
+};
+    
+struct rooted_mon {
+    svector<lpvar>   m_vars;
+    vector<index_with_sign>  m_mons; // canonize_monomial(m_mons[j]) gives m_vector_of_rooted_monomials[m_i]
+
+    rooted_mon(const svector<lpvar>& vars, unsigned i, const rational& sign): m_vars(vars) {
+        SASSERT(lp::is_non_decreasing(vars));
+        push_back(index_with_sign(i, sign));
+    }
+    lpvar operator[](unsigned k) const { return m_vars[k];}
+    unsigned size() const { return m_vars.size(); }
+    unsigned orig_index() const { return m_mons.begin()->m_i; }
+    const rational& orig_sign() const { return m_mons.begin()->m_sign; }
+    const index_with_sign& orig() const { return *m_mons.begin(); }
+    const svector<lpvar> & vars() const {return m_vars;}
+    void push_back(const index_with_sign& ind_s) {
+        m_mons.push_back(ind_s);
+    }
+};
+
+
+struct rooted_mon_table {
+    std::unordered_map<svector<lpvar>,
+                       unsigned, // points to vec()
+                       hash_svector>                                 m_map;
+    vector<rooted_mon>                                               m_vector;
+    // for every k 
+    // for each i in m_rooted_monomials_containing_var[k]
+    // m_vector_of_rooted_monomials[i] contains k
+    std::unordered_map<lpvar, std::unordered_set<unsigned>>          m_mons_containing_var;
+
+    // A map from m_vector_of_rooted_monomials to a set
+    // of sets of m_vector_of_rooted_monomials,
+    // such that for every i and every h in m_proper_factors[i] we have m_vector[i] as a proper factor of m_vector[h]
+    std::unordered_map<unsigned, std::unordered_set<unsigned>>       m_proper_factors;
+    // points to m_vector
+    svector<unsigned>                                                m_to_refine;
+    // maps the indices of the regular monomials to the rooted monomial indices
+    std::unordered_map<unsigned, index_with_sign>                    m_reg_to_rm;
+
+    void print_stats(std::ostream& out) const {
+        static double ratio = 1;
+        double s = 0;
+        for (const auto& p : m_map) {
+            s += m_vector[p.second].m_mons.size();
+        }
+        double r = s /m_map.size();
+        if (r > ratio) {
+            ratio = r;
+            out << "rooted mons "  << m_map.size() << ", ratio = " << r << "\n";
+        }
+    }
+
+    const svector<unsigned>& to_refine() { return m_to_refine; }
+
+    bool list_is_consistent(const vector<index_with_sign>& list, const std::unordered_set<unsigned>& to_refine_reg) const {
```

Comment:
work with empty list?
for (unsigned i = 0, j = 0; i < list.size() && (j == 0 || i == j); ++i) {
      if (to_refine_reg.contains(list[i].index())) j++;
}
return j == 0 || j == list.size();